### PR TITLE
A support plugin for lightningd's invoice and pay API call. 

### PR DESCRIPTION
**Pull request by renepickhardt**
If `lightningd` users want to make AMP payments `lightningd` could use this plugin to compute the optimal split off the amount that helps to rebalance it's own channels in the best way.

you can test the code by running `example.py` before you need to do `lightning-cli listfunds > listfunds.json` as described in `example.py`

Looking forward for your review and thoughts.

edit (for easier reference some background info):
More theory about imbalances and the algorithm to decrease the imblance of a node was
suggested by this research by Rene Pickhardt and Mariusz Nowostawski: https://arxiv.org/abs/1912.09555

This software is an adopted version as discussed in this post on the lightning-dev mailing
list: https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-January/002418.html
and was suggested to be created by ZmnSCPxj.

### CODE REVIEW

File: balanced_amp_payments/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+The plugin exposes no new command line options:
+   
+## JSON-RPC methods
+
+The plugin also exposes the following methods:
+
+ - `amp_invoice_amounts`: splits an amount that is supposed to be received into smaller amounts respecing in a way that best decreases the imbalance of the nodes channels
+
+ - `amp_pay_amounts`: splits an amount that is supposed to be send into smaller amounts respecing in a way that best decreases the imbalance of the nodes channels
```
[the remark this answers is not in the document]
Comment (renepickhardt) [reply balanced_amp_payments/README.md]: the research results but I agree as a documentation this is not well written. I should say in a way that maximally reduces the imbalance of channels / balances

File: balanced_amp_payments/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Balanced AMP Payments
+
+This plugin computes an optimal split of a payment amount for the use of AMP.
+The split is optimal in the sense that it reduces the imbalance of the funds of the node.
+
+More theory about imbalances and the algorithm to decrease the imblance of a node was
+suggested by this research by Rene Pickhardt and Mariusz Nowostawski: https://arxiv.org/abs/1912.09555
+
+This software is an adopted version as discussed in this post on the lightning-dev mailing
+list: https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-January/002418.html
```
[the remark this answers is not in the document]
Comment (renepickhardt) [reply balanced_amp_payments/README.md]: sure. I usually like explicit urls but I have no strong feelings about that

**Comment by cdecker** [comment]
Very interesting idea, we briefly discussed this in person, but would you be open to adding functionality to perform a payment if I build a class that extracts the common pieces of a `pay` call? It'd be a class representing a payment, and it'd then manage calls to `getroute`, splitting the payment into partial payments, and reattempt failed payments, with the necessary entrypoints to add your `getroute` specialization.

**Comment by renepickhardt** [comment]
sure! Though i am not completely shure how the API for for the Pay class would look like. But I guess I will be able to read the code and hang this plugin in. 

On a side note: It would also be interesting to find a tipping point between necessary fees and desired reduction in imbalance.

**Comment by chrisguida** [comment]
This plugin has been archived here https://github.com/lightningd/plugins/blob/master/archived.md#archived-plugins

Feel free to reopen if anyone wants to continue working on it!

**Pull request closed without merging**